packages/catlog/src/tt/notebook_elab.rs

```diff
@@ -0,0 +1,332 @@
+                synthesized: synthed_dom_ty,
+            })
+        } else {
+            Some(())
+        };
+        let correct_cod = if synthed_cod_ty != cod_ty {
+            self.error(Error::MismatchedCodTypes {
+                expected: cod_ty,
+                synthesized: synthed_cod_ty,
+            })
+        } else {
+            Some(())
+        };
+        let _ = (correct_dom?, correct_cod?);
+        Some((
+            name,
+            label,
+            TyS::morphism(mor_type.clone(), dom_s, cod_s),
+            TyV::morphism(mor_type, dom_v, cod_v),
+        ))
+    }
+
+    fn notebook(&mut self, cells: &[(Uuid, &ModelJudgment)]) -> Option<(TyS, TyV)> {
+        let mut field_ty0s = Vec::new();
+        let mut field_ty_vs = Vec::new();
+        let self_var = self.intro(text_seg("self"), label_seg("self"), None).as_neu();
+        let c = self.checkpoint();
+        for (id, cell) in cells.iter() {
+            self.current_cell = Some(*id);
+            let (name, label, ty_s, ty_v) = match cell {
+                ModelJudgment::Object(ob_decl) => self.object_cell(ob_decl),
+                ModelJudgment::Morphism(mor_decl) => self.morphism_cell(mor_decl),
+            }?;
+            field_ty0s.push((name, (label, ty_v.ty0())));
+            field_ty_vs.push((name, (label, ty_v.clone())));
+            self.ctx.scope.push((name, label, Some(ty_v.clone())));
+            self.ctx.env =
+                self.ctx.env.snoc(TmV::Neu(TmN::proj(self_var.clone(), name, label), ty_v));
+        }
+        self.reset_to(c);
+        let field_tys: Row<_> = field_ty_vs
+            .iter()
+            .map(|(name, (label, ty_v))| (*name, (*label, self.evaluator().quote_ty(ty_v))))
+            .collect();
+        let field_ty0s: Row<_> = field_ty0s.into_iter().collect();
+        let r_s = RecordS::new(field_ty0s.clone(), field_tys.clone());
+        let r_v = RecordV::new(field_ty0s, self.ctx.env.clone(), field_tys, Dtry::empty());
+        Some((TyS::record(r_s), TyV::record(r_v)))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use notebook_types::v1::{Document, ModelDocumentContent, ModelJudgment, Notebook};
+    use serde_json;
+    use std::fmt::Write;
+    use std::{fs, rc::Rc};
+
+    use expect_test::{Expect, expect};
+
+    use crate::{
+        stdlib::th_schema,
+        tt::{
+            modelgen::{generate, model_output},
+            notebook_elab::Elaborator,
+            toplevel::Toplevel,
+        },
+    };
+
+    fn elab_example(name: &str, expected: Expect) {
+        let src = fs::read_to_string(format!("examples/{name}.json")).unwrap();
+        let doc: ModelDocumentContent = serde_json::from_str(&src).unwrap();
+        let cells = doc
+            .notebook
+            .cells()
+            .filter_map(|c| match c {
+                notebook_types::v1::NotebookCell::Formal { id, content } => Some((*id, content)),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
```

Comment:
Should the notebook elaborator have an entry point that does this filtering? We'll need to do the same thing when processing a notebook from the frontend.

Reply:
I think that the frontend should do the filtering, because then it will be tracked by the solid incremental stuff?